### DISCUSSION TITLE
What's the reason for adding pos_embed and token_embed?

### DISCUSSION THREAD

**Discussion by shi-yan** [karpathy/build-nanogpt#70] [Q&A]
It seems strange that the token embedding is added to the positional embedding because addition can obscure the original values. For example, with 1+4=5, knowing only 5 makes it difficult to determine the original values. Why is addition used?

A more logical approach might be concatenation, which preserves the original values but increases computation. But why must the positional embedding be the same length as the token embedding, we can potentially use a shorter positional embedding?

The GPT-2 implementation video appears to use a trained positional embedding rather than the cos/sin functions described in the attention paper. Why is this approach taken?

**Comment by shayanjl** [comment]
Addition is used mainly because it keeps the representation size fixed. The transformer expects one vector per token position with a constant hidden size, so adding token and positional embeddings gives a single vector of the same dimension that can go directly into the model. Concatenation would double or otherwise increase the width, which would increase parameters and compute in all following layers. In practice, the model can still learn to separate “content” and “position” information inside that shared space, so the addition is not as destructive as the simple `1+4=5` analogy suggests.

The positional embedding is usually the same size as the token embedding for this reason: both must live in the same hidden space if you want to add them. A shorter positional embedding would require an extra projection step anyway, so the simplicity advantage mostly disappears.

As for learned positional embeddings versus sinusoidal ones, both work. Sinusoidal embeddings were used in the original Transformer because they are simple and can extrapolate to longer positions more naturally. GPT-2 used learned positional embeddings because they are flexible, easy to train, and worked well in practice for a fixed context length. So the short answer is: addition is mostly for efficiency and architectural simplicity, and learned positional embeddings are a practical design choice rather than a requirement.
